### ISSUE TITLE
Question: EventingBasicConsumer with Manual Ack and IModel

### ISSUE THREAD

**Issue by Nicholi** [rabbitmq/rabbitmq-dotnet-client#145]
This is mostly a question concerning the use of EventingBasicConsumer with manual acknowledgement of messages. I've read countless times the IConnection of RabbitMQ is thread safe, however IModel is not. I haven't seen this addressed exactly for EventingBasicConsumer and its callbacks.

In a situation of an IConnection, IModel, and EventingBasicConsumer initiated on a single thread is there any problem with accessing the IModel in the Received callback to acknowledge messages? Since each Received will be executed on whatever thread handles the callback (all I know is it's not the same thread).

Is this fine?
```csharp
	var consumer = new EventingBasicConsumer(channel);
	consumer.Received += (o, args) =>
		{
			var consumerObj = (EventingBasicConsumer)o;
			var props = args.BasicProperties;
			var deliveryTag = args.DeliveryTag;
			
			// do stuff with message/etc
			
			consumerObj.Model.BasicAck(deliveryTag, false);
			// or
			// consumerObj.Model.BasicNack(deliveryTag, false, true);
			// or 
			// consumerObj.Model.BasicCancel(consumer.ConsumerTag);
		};
```

Or do we need this to be inherently thread-safe?
```csharp
	var consumer = new EventingBasicConsumer(channel);
	consumer.Received += (o, args) =>
		{
			var consumerObj = (EventingBasicConsumer)o;
			var props = args.BasicProperties;
			var deliveryTag = args.DeliveryTag;
			
			// do stuff with message/etc
			
			lock (consumerObj.Model)
			{
				consumerObj.Model.BasicAck(deliveryTag, false);
			}
		};
```

In other words do we need to lock on the IModel before acknowledging? Or in this scenario of simply consuming messages with EventingBasicConsumer (in which the only reason multiple threads are involved because of how the Received events are processed) it's just fine?

**Comment by michaelklishin** [comment]
Please ask questions on rabbitmq-users.

"Thread safe" makes no sense unless you also specify "in what context". The context, almost certainly, is concurrent publishing, not consuming.

You can — in fact, you must, because delivery tags are scoped per channel — ack on the same channel (`IModel`) that you got the delivery on.

**Comment by manigandham** [comment]
Hi @michaelklishin 

I believe this would be a possible issue with the driver itself. The .NET guide shows that `IModel` is not thread safe: https://www.rabbitmq.com/dotnet-api-guide.html

> Sharing Channels Between Threads

> As a rule of thumb, `IModel` instances should not be used by more than one thread simultaneously: application code should maintain a clear notion of thread ownership for IModel instances. If more than one thread needs to access a particular IModel instances, the application should enforce mutual exclusion itself. One way of achieving this is for all users of an IModel to lock the instance itself:
> 
> ```
> IModel ch = RetrieveSomeSharedIModelInstance();
> lock (ch) {
>   ch.BasicPublish(...);
> }
> ```

The guide also shows this example for 

> ```
> var consumer = new EventingBasicConsumer(channel);
> consumer.Received += (ch, ea) =>
>                 {
>                     var body = ea.Body;
>                     // ... process the message
>                     ch.BasicAck(ea.DeliveryTag, false);
>                 };
> ```

Since the `EventingBasicConsumer` uses a event model, the dispatched events can be processed in any thread and the ACKs are being sent using a reference to the `IModel` sent through the event arguments. If `IModel` isn't thread-safe, then (I'm assuming) there should be a `lock` statement when calling the BasicAck method as well, just like BasicPublish.

Is this by design or is there potential to make the `IModel` interface thread-safe and remove the need for `lock` statements?

**Comment by Nicholi** [comment]
I am guessing IModel only isn't safe for concurrent publishing, that's why my question was pretty much ignored. If we are simply ACKing consumers that must be different?

I'd like a clean answer as well because from reading the docs I came to the same conclusion as you. I have implemented without a lock, and the responses are being handled async on different threads, with no problems. Though that doesn't necessarily mean anything.

**Comment by manigandham** [comment]
There are some threads in the rabbitmq-users google group

1. https://groups.google.com/forum/#!searchin/rabbitmq-users/thread-safe/rabbitmq-users/KfIqEbdAKws/xZ_Ei6SFBWAJ

1. https://groups.google.com/forum/#!searchin/rabbitmq-users/imodel$20thread/rabbitmq-users/2TEusuFq_n8/ReSs-bjwxKoJ

The information is still confusing. BasicConsume/BasicCancel seem to be non-threadsafe as well but just not expected to be called concurrently, which is probably fine, however ACKs are still information being transmitted so I'm trying to understand why BasicPublish would require locks but BasicAcks would not.

The general advice seems to be to use a `IModel` per thread but that's not really compatible with the EventingBasicConsumer. There seems to be some lower-level locking happening anyway within these methods so maybe the explicit lock isn't required but either way it would be nice to have a clear answer on this.

If anything, it would be best to make `IModel` thread-safe to avoid all these issues and have the driver take care of the locking work.

**Comment by michaelklishin** [comment]
Because basic.publish is at least 2, and typically 3 or more, frames on the wire, so you can get incorrect interleaving.

Acking with multiple set to true on a shared channel is also unsafe, one-by-one can work fine. Concurrent consuming is fine unless you register consumers concurrently.

Questions beling to rabbitmq-users.

**Comment by michaelklishin** [comment]
Very often eventing consumer handles deliveries on the very same thread that its channel was opened on. If you redispatch drliveries it is up to you to synchronize. This is neither a problem in the library nor with the example: we cannot produce examples that are universally safe so we simplify them.

**Comment by manigandham** [comment]
> This is neither a problem in the library nor with the example: we cannot produce examples that are universally safe so we simplify them.

Doesn't making `IModel` thread-safe solve this? If the current implementation is by design, then this isn't pointing out a problem or just asking questions but rather suggesting an improvement to the driver.

The java client does have a thread safe channel interface that serializes commands which can help make application code simpler and safer for publishing.

**Comment by michaelklishin** [comment]
Java client is no different in this regard: concurrent publishing on a shared channel or acking with multi = true is asking for trouble.

The docs do mention that channels should not be shared between channels.

**Comment by nathan-alden** [comment]
I have exactly the same question. I wrote some common library code that sets up asynchronous receipt of messages by using the `EventingBasicConsumer.Received` event. It is not clear to me how, if at all, I should be synchronizing the event handler's access to `IModel`. This thread didn't appear to answer the OP's question. Does anyone know what the current guidance is?

**Comment by michaelklishin** [comment]
Please post questions to rabbitmq-users or Stack Overflow. RabbitMQ uses GitHub issues for specific actionable items engineers can work on, not questions. Thank you.

**Comment by michaelklishin** [comment]
Consumer instances can use a synchronized collection or perform any other form of synchronization. Unless you publish on an `IModel` instance, it can be safe to share it.

It's usually easy to make a consumer aware of its channel (`IModel`), which opens a number of other possibilities.

Lastly, if there aren't many consumers you can give its own a separate channel and potentially side step the problem entirely.

**Comment by FarhadJabiyev** [comment]
> Please post questions to rabbitmq-users or Stack Overflow. RabbitMQ uses GitHub issues for specific actionable items engineers can work on, not questions. Thank you.
I asked question in So, but no any answer from RabbitMq team yet. May be they are not aware even about that case as well. https://stackoverflow.com/questions/59125143/is-manual-acknowledgment-of-imodel-thread-safe-in-rabbitmq-net-client

**Comment by michaelklishin** [comment]
@FarhadJabiyev you are not entitled to a response from the core team. There are orders of magnitude more questions than engineers on our team.

**Comment by michaelklishin** [comment]
I have responded [in this StackOverflow thread](https://stackoverflow.com/questions/59125143/is-manual-acknowledgment-of-imodel-thread-safe-in-rabbitmq-net-client).
